docs/DifferentiableProgrammingImplementation.md

```diff
@@ -0,0 +1,1012 @@
+
+Finally, the JVP returns a tuple of the original result and the generated differential function.
+
+```swift
+// JVP: replaces all function applications with JVP applications.
+sil @jvp_foo : $(Float) -> (Float, (Float) -> Float) {
+bb0(%x):
+  (%y1, %df_sin) = apply @jvp_sin(%x)
+  (%y2, %df_cos) = apply @jvp_cos(%x)
+  (%y3, %df_mul) = apply @jvp_mul(%y1, %y2)
+  // Return tuple of original result and differential.
+  return (%y3, { %dx in
+    %dy1 = apply %df_sin(%dx)
+    %dy2 = apply %df_cos(%dx)
+    %dy3 = apply %df_mul(%dy1, %dy2)
+    return %dy3
+  })
+}
+```
+
+### Visiting the original function to create a JVP
+
+In the Swift compiler, we have a class called [`JVPEmitter`](https://github.com/apple/swift/blob/tensorflow/lib/SILOptimizer/Mandatory/Differentiation.cpp) which subclasses [`TypeSubstCloner`](https://github.com/apple/swift/blob/master/include/swift/SIL/TypeSubstCloner.h). This uses the [visitor pattern](https://en.wikipedia.org/wiki/Visitor_pattern) to visit the original function and generate the JVP function. Taking a look at the `JVPEmitter` class, there are methods like `visitApplyInst`, `visitStructExtractInst`, etc. Each of these methods visit an instruction that is important in the generation of the JVP and differential, and emits a newly mapped version. We handle each type of instruction differently, explained below.
+
+One important note is that in the JVP, we visit every single instruction in the original function - sometimes it’s an exact copy, and other times there is some special logic we wrote to handle it differently (e.g. like control-flow discussed below). This is so that the JVP function behaves just like the original function. With this, if the original function has a print statement, the JVP will as well. However, when we consider the differential, we will only transform SIL instructions from the original that we deem to be fit (so no print statement in the differential!). These instructions are those that should be differentiated, which uses the activity analysis calculated earlier to determine which SIL instructions are important in computing the tangent values of a function. 
+
+In Swift code, the generated differential function can be written as a closure, capturing the “callee differentials” from the JVP. But in SIL, all functions are top-level; closures are represented as top-level functions with captured values as an explicit argument. This requires a differential struct which will be discussed in the next section. What’s important here is that the `JVPEmitter` emits code both in a top level JVP function, but also a corresponding top level differential function.
```

Comment:
Again, this does not match the current state.

```suggestion
In Swift code, the generated differential function can be written as a closure, capturing the "callee differentials" from the JVP. But in SIL, all functions are top-level; closures are represented as top-level functions with captured values as an explicit argument. This requires a differential struct which will be discussed in the next section. What’s important here is that the `JVPEmitter` emits code both in a top level JVP function, but also a corresponding top level differential function.
```

Reply:
What changes must be made to this text then? (aside from the quotation correction)

Reply:
There is no `JVPEmitter`, so what actually generates that?  This largely impacts the last sentence.

Reply:
@dan-zheng could you patch this sentence since you know the most about this?

Reply:
It's been renamed to `JVPCloner`.

Reply:
```suggestion
In Swift code, the generated differential function can be written as a closure, capturing the "callee differentials" from the JVP. But in SIL, all functions are top-level; closures are represented as top-level functions with captured values as an explicit argument. This requires a differential struct which will be discussed in the next section. What’s important here is that the `JVPCloner` emits code both in a top level JVP function, but also a corresponding top level differential function.
```